Prometheus panic: &runtime.TypeAssertionError

**What did you do?**
Running Prometheus inside Kubernetes. When I scale down number of replicas for a deployment, Prometheus panic.

**What did you expect to see?**
Not panic

**What did you see instead? Under which circumstances?**
When I scale 350 replicas to 0, Prometheus crashed.

**Environment**

* System information:

        CentOS Linux release 7.2.1511 (Core)
	Linux 3.10.0-327.4.4.el7.x86_64 x86_64

* Prometheus version: 

       Docker image: prom/prometheus:v1.3.1

* Docker version:

       Server:
        Version:      1.11.2
        API version:  1.23
        Go version:   go1.5.4
        Git commit:   b9f10c9
        Built:        Wed Jun  1 21:23:11 2016
        OS/Arch:      linux/amd64

* Kubelet version:

       Kubernetes v1.4.1

* Prometheus configuration file:
```
global:
  scrape_interval: 600s

scrape_configs:
  - job_name: 'cadvisor-exporter'
    scrape_interval: 60s
    scheme: 'http'
    kubernetes_sd_configs:
    - role: node
      api_server: https://API_SERVER
      tls_config:
        ca_file: CA_CRT_PATH
    relabel_configs:
      - source_labels: [__meta_kubernetes_node_address_InternalIP]
        regex: (.*)
        replacement: ${1}:4194
        action: replace
        target_label: __address__
      - source_labels: [instance]
        regex: (.*)\.(.*)\.(.*)
        replacement: ${1}
        action: replace
        target_label: hostname
  - job_name: 'node-exporter'
    scrape_interval: 60s
    scheme: 'http'
    kubernetes_sd_configs:
    - role: node
      api_server: https://API_SERVER
      tls_config:
        ca_file: CA_CRT_PATH
    relabel_configs:
      - source_labels: [__meta_kubernetes_node_address_InternalIP]
        regex: (.*)
        replacement: ${1}:9100
        action: replace
        target_label: __address__
      - source_labels: [instance]
        regex: (.*)\.(.*)\.(.*)
        replacement: ${1}
        action: replace
        target_label: hostname
      - action: labelmap
        regex: __meta_kubernetes_node_label_(.+)
  - job_name: 'prod-pod-exporter'
    scrape_interval: 60s
    scheme: 'http'
    kubernetes_sd_configs:
    - role: pod
      api_server: https://API_SERVER
      tls_config:
        ca_file: CA_CRT_PATH
    tls_config:
      ca_file: CA_CRT_PATH
      server_name: '*.k8s.local'
    relabel_configs:
    - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
      action: replace
      target_label: __metrics_path__
      regex: (.+)
    - source_labels: [__meta_kubernetes_pod_label_prometheus_io_scrape]
      action: keep
      regex: true
    - source_labels: [__meta_kubernetes_pod_container_port_name]
      action: keep
      regex: prometheus-port
    - source_labels: [__address__, __meta_kubernetes_pod_container_port_map_prometheus_port]
      action: replace
      regex: (.+):(?:\d+);(\d+)
      replacement: ${1}:${2}
      target_label: __address__
    - action: labelmap
      regex: __meta_kubernetes_pod_label_(.+)
    - source_labels: [__meta_kubernetes_pod_namespace]
      action: replace
      target_label: kubernetes_namespace
    - source_labels: [__meta_kubernetes_pod_name]
      action: replace
      target_label: kubernetes_pod_name
```

* Logs:
```
E1109 20:02:54.793285       1 runtime.go:64] Observed a panic: &runtime.TypeAssertionError{interfaceString:"interface {}", concreteString:"cache.DeletedFinalStateUnknown", assertedString:"*v1.Pod", missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod)
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/pkg/util/runtime/runtime.go:70
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/pkg/util/runtime/runtime.go:63
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:479
/usr/local/go/src/runtime/panic.go:458
/usr/local/go/src/runtime/iface.go:201
/go/src/github.com/prometheus/prometheus/retrieval/discovery/kubernetes/pod.go:77
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache/controller.go:182
<autogenerated>:52
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache/shared_informer.go:408
/usr/local/go/src/runtime/asm_amd64.s:2086
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod [recovered]
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod

goroutine 419 [running]:
  panic(0x17a4060, 0xc514b803c0)
/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/pkg/util/runtime/runtime.go:56 +0x126
panic(0x17a4060, 0xc514b803c0)
/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/prometheus/prometheus/retrieval/discovery/kubernetes.(*Pod).Run.func3(0x1816540, 0xc502773820)
/go/src/github.com/prometheus/prometheus/retrieval/discovery/kubernetes/pod.go:77 +0x166
github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache.ResourceEventHandlerFuncs.OnDelete(0xc4d5821240, 0xc4d5821280, 0xc5266dcf40, 0x1816540, 0xc502773820)
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache/controller.go:182 +0x49
github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache.(*ResourceEventHandlerFuncs).OnDelete(0xc4d58212a0, 0x1816540, 0xc502773820)
<autogenerated>:52 +0x78
github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache.(*processorListener).run(0xc4cbe56e80, 0xc49c814ba0)
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache/shared_informer.go:408 +0x356
created by github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache.(*sharedIndexInformer).AddEventHandler
/go/src/github.com/prometheus/prometheus/vendor/k8s.io/client-go/1.5/tools/cache/shared_informer.go:257 +0x24b
```

--- a/retrieval/discovery/kubernetes/endpoints.go
+++ b/retrieval/discovery/kubernetes/endpoints.go
@@ -83,31 +83,38 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 
 	e.endpointsInf.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(o interface{}) {
-			send(e.buildEndpoints(o.(*apiv1.Endpoints)))
+			eps, err := convertToEndpoints(o)
+			if err != nil {
+				e.logger.With("err", err).Errorln("converting to Endpoints object failed")
+				return
+			}
+			send(e.buildEndpoints(eps))
 		},
 		UpdateFunc: func(_, o interface{}) {
-			send(e.buildEndpoints(o.(*apiv1.Endpoints)))
+			eps, err := convertToEndpoints(o)
+			if err != nil {
+				e.logger.With("err", err).Errorln("converting to Endpoints object failed")
+				return
+			}
+			send(e.buildEndpoints(eps))
 		},
 		DeleteFunc: func(o interface{}) {
-			endpoints, isEndpoints := o.(*apiv1.Endpoints)
-			if !isEndpoints {
-				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					e.logger.Errorln("Received unexpected object: %v", o)
-					return
-				}
-				endpoints, ok = deletedState.Obj.(*apiv1.Endpoints)
-				if !ok {
-					e.logger.Errorln("DeletedFinalStateUnknown contained non-Endpoints object: %v", deletedState.Obj)
-					return
-				}
+			eps, err := convertToEndpoints(o)
+			if err != nil {
+				e.logger.With("err", err).Errorln("converting to Endpoints object failed")
+				return
 			}
-
-			send(&config.TargetGroup{Source: endpointsSource(endpoints.ObjectMeta)})
+			send(&config.TargetGroup{Source: endpointsSource(eps)})
 		},
 	})
 
-	serviceUpdate := func(svc *apiv1.Service) {
+	serviceUpdate := func(o interface{}) {
+		svc, err := convertToService(o)
+		if err != nil {
+			e.logger.With("err", err).Errorln("converting to Service object failed")
+			return
+		}
+
 		ep := &apiv1.Endpoints{}
 		ep.Namespace = svc.Namespace
 		ep.Name = svc.Name
@@ -122,33 +129,33 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 	e.serviceInf.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		// TODO(fabxc): potentially remove add and delete event handlers. Those should
 		// be triggered via the endpoint handlers already.
-		AddFunc:    func(o interface{}) { serviceUpdate(o.(*apiv1.Service)) },
-		UpdateFunc: func(_, o interface{}) { serviceUpdate(o.(*apiv1.Service)) },
-		DeleteFunc: func(o interface{}) {
-			service, isService := o.(*apiv1.Service)
-			if !isService {
-				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					e.logger.Errorln("Received unexpected object: %v", o)
-					return
-				}
-				service, ok = deletedState.Obj.(*apiv1.Service)
-				if !ok {
-					e.logger.Errorln("DeletedFinalStateUnknown contained non-Service object: %v", deletedState.Obj)
-					return
-				}
-			}
-
-			serviceUpdate(service)
-		},
+		AddFunc:    func(o interface{}) { serviceUpdate(o) },
+		UpdateFunc: func(_, o interface{}) { serviceUpdate(o) },
+		DeleteFunc: func(o interface{}) { serviceUpdate(o) },
 	})
 
 	// Block until the target provider is explicitly canceled.
 	<-ctx.Done()
 }
 
-func endpointsSource(ep apiv1.ObjectMeta) string {
-	return "endpoints/" + ep.Namespace + "/" + ep.Name
+func convertToEndpoints(o interface{}) (*apiv1.Endpoints, error) {
+	endpoints, isEndpoints := o.(*apiv1.Endpoints)
+	if !isEndpoints {
+		deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("Received unexpected object: %v", o)
+		}
+		endpoints, ok = deletedState.Obj.(*apiv1.Endpoints)
+		if !ok {
+			return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Endpoints object: %v", deletedState.Obj)
+		}
+	}
+
+	return endpoints, nil
+}
+
+func endpointsSource(ep *apiv1.Endpoints) string {
+	return "endpoints/" + ep.ObjectMeta.Namespace + "/" + ep.ObjectMeta.Name
 }
 
 const (
@@ -164,7 +171,7 @@ func (e *Endpoints) buildEndpoints(eps *apiv1.Endpoints) *config.TargetGroup {
 	}
 
 	tg := &config.TargetGroup{
-		Source: endpointsSource(eps.ObjectMeta),
+		Source: endpointsSource(eps),
 	}
 	tg.Labels = model.LabelSet{
 		namespaceLabel:     lv(eps.Namespace),

--- a/retrieval/discovery/kubernetes/endpoints.go
+++ b/retrieval/discovery/kubernetes/endpoints.go
@@ -89,7 +89,21 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 			send(e.buildEndpoints(o.(*apiv1.Endpoints)))
 		},
 		DeleteFunc: func(o interface{}) {
-			send(&config.TargetGroup{Source: endpointsSource(o.(*apiv1.Endpoints).ObjectMeta)})
+			endpoints, isEndpoints := o.(*apiv1.Endpoints)
+			if !isEndpoints {
+				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					e.logger.Errorln("Received unexpected object: %v", o)
+					return
+				}
+				endpoints, ok = deletedState.Obj.(*apiv1.Endpoints)
+				if !ok {
+					e.logger.Errorln("DeletedFinalStateUnknown contained non-Endpoints object: %v", deletedState.Obj)
+					return
+				}
+			}
+
+			send(&config.TargetGroup{Source: endpointsSource(endpoints.ObjectMeta)})
 		},
 	})
 
@@ -110,7 +124,23 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 		// be triggered via the endpoint handlers already.
 		AddFunc:    func(o interface{}) { serviceUpdate(o.(*apiv1.Service)) },
 		UpdateFunc: func(_, o interface{}) { serviceUpdate(o.(*apiv1.Service)) },
-		DeleteFunc: func(o interface{}) { serviceUpdate(o.(*apiv1.Service)) },
+		DeleteFunc: func(o interface{}) {
+			service, isService := o.(*apiv1.Service)
+			if !isService {
+				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					e.logger.Errorln("Received unexpected object: %v", o)
+					return
+				}
+				service, ok = deletedState.Obj.(*apiv1.Service)
+				if !ok {
+					e.logger.Errorln("DeletedFinalStateUnknown contained non-Service object: %v", deletedState.Obj)
+					return
+				}
+			}
+
+			serviceUpdate(service)
+		},
 	})
 
 	// Block until the target provider is explicitly canceled.

--- a/retrieval/discovery/kubernetes/endpoints_test.go
+++ b/retrieval/discovery/kubernetes/endpoints_test.go
@@ -20,6 +20,7 @@ import (
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
 	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/tools/cache"
 )
 
 func endpointsStoreKeyFunc(obj interface{}) (string, error) {
@@ -240,6 +241,21 @@ func TestEndpointsDiscoveryDelete(t *testing.T) {
 	k8sDiscoveryTest{
 		discovery:  n,
 		afterStart: func() { go func() { eps.Delete(makeEndpoints()) }() },
+		expectedRes: []*config.TargetGroup{
+			&config.TargetGroup{
+				Source: "endpoints/default/testendpoints",
+			},
+		},
+	}.Run(t)
+}
+
+func TestEndpointsDiscoveryDeleteUnknownCacheState(t *testing.T) {
+	n, _, eps, _ := makeTestEndpointsDiscovery()
+	eps.GetStore().Add(makeEndpoints())
+
+	k8sDiscoveryTest{
+		discovery:  n,
+		afterStart: func() { go func() { eps.Delete(cache.DeletedFinalStateUnknown{Obj: makeEndpoints()}) }() },
 		expectedRes: []*config.TargetGroup{
 			&config.TargetGroup{
 				Source: "endpoints/default/testendpoints",

--- a/retrieval/discovery/kubernetes/node.go
+++ b/retrieval/discovery/kubernetes/node.go
@@ -66,7 +66,21 @@ func (n *Node) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 			send(n.buildNode(o.(*apiv1.Node)))
 		},
 		DeleteFunc: func(o interface{}) {
-			send(&config.TargetGroup{Source: nodeSource(o.(*apiv1.Node))})
+			node, isNode := o.(*apiv1.Node)
+			if !isNode {
+				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					n.logger.Errorln("Received unexpected object: %v", o)
+					return
+				}
+				node, ok = deletedState.Obj.(*apiv1.Node)
+				if !ok {
+					n.logger.Errorln("DeletedFinalStateUnknown contained non-Node object: %v", deletedState.Obj)
+					return
+				}
+			}
+
+			send(&config.TargetGroup{Source: nodeSource(node)})
 		},
 		UpdateFunc: func(_, o interface{}) {
 			send(n.buildNode(o.(*apiv1.Node)))

--- a/retrieval/discovery/kubernetes/node.go
+++ b/retrieval/discovery/kubernetes/node.go
@@ -63,32 +63,49 @@ func (n *Node) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 	}
 	n.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(o interface{}) {
-			send(n.buildNode(o.(*apiv1.Node)))
+			node, err := convertToNode(o)
+			if err != nil {
+				n.logger.With("err", err).Errorln("converting to Node object failed")
+				return
+			}
+			send(n.buildNode(node))
 		},
 		DeleteFunc: func(o interface{}) {
-			node, isNode := o.(*apiv1.Node)
-			if !isNode {
-				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					n.logger.Errorln("Received unexpected object: %v", o)
-					return
-				}
-				node, ok = deletedState.Obj.(*apiv1.Node)
-				if !ok {
-					n.logger.Errorln("DeletedFinalStateUnknown contained non-Node object: %v", deletedState.Obj)
-					return
-				}
+			node, err := convertToNode(o)
+			if err != nil {
+				n.logger.With("err", err).Errorln("converting to Node object failed")
+				return
 			}
-
 			send(&config.TargetGroup{Source: nodeSource(node)})
 		},
 		UpdateFunc: func(_, o interface{}) {
-			send(n.buildNode(o.(*apiv1.Node)))
+			node, err := convertToNode(o)
+			if err != nil {
+				n.logger.With("err", err).Errorln("converting to Node object failed")
+				return
+			}
+			send(n.buildNode(node))
 		},
 	})
 
 	// Block until the target provider is explicitly canceled.
 	<-ctx.Done()
+}
+
+func convertToNode(o interface{}) (*apiv1.Node, error) {
+	node, isNode := o.(*apiv1.Node)
+	if !isNode {
+		deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("Received unexpected object: %v", o)
+		}
+		node, ok = deletedState.Obj.(*apiv1.Node)
+		if !ok {
+			return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Node object: %v", deletedState.Obj)
+		}
+	}
+
+	return node, nil
 }
 
 func nodeSource(n *apiv1.Node) string {

--- a/retrieval/discovery/kubernetes/node_test.go
+++ b/retrieval/discovery/kubernetes/node_test.go
@@ -270,6 +270,36 @@ func TestNodeDiscoveryDelete(t *testing.T) {
 	}.Run(t)
 }
 
+func TestNodeDiscoveryDeleteUnknownCacheState(t *testing.T) {
+	n, i := makeTestNodeDiscovery()
+	i.GetStore().Add(makeEnumeratedNode(0))
+
+	k8sDiscoveryTest{
+		discovery:  n,
+		afterStart: func() { go func() { i.Delete(cache.DeletedFinalStateUnknown{Obj: makeEnumeratedNode(0)}) }() },
+		expectedInitial: []*config.TargetGroup{
+			&config.TargetGroup{
+				Targets: []model.LabelSet{
+					model.LabelSet{
+						"__address__": "1.2.3.4:10250",
+						"instance":    "test0",
+						"__meta_kubernetes_node_address_InternalIP": "1.2.3.4",
+					},
+				},
+				Labels: model.LabelSet{
+					"__meta_kubernetes_node_name": "test0",
+				},
+				Source: "node/test0",
+			},
+		},
+		expectedRes: []*config.TargetGroup{
+			&config.TargetGroup{
+				Source: "node/test0",
+			},
+		},
+	}.Run(t)
+}
+
 func TestNodeDiscoveryUpdate(t *testing.T) {
 	n, i := makeTestNodeDiscovery()
 	i.GetStore().Add(makeEnumeratedNode(0))

--- a/retrieval/discovery/kubernetes/pod.go
+++ b/retrieval/discovery/kubernetes/pod.go
@@ -71,32 +71,49 @@ func (p *Pod) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 	}
 	p.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(o interface{}) {
-			send(p.buildPod(o.(*apiv1.Pod)))
+			pod, err := convertToPod(o)
+			if err != nil {
+				p.logger.With("err", err).Errorln("converting to Pod object failed")
+				return
+			}
+			send(p.buildPod(pod))
 		},
 		DeleteFunc: func(o interface{}) {
-			pod, isPod := o.(*apiv1.Pod)
-			if !isPod {
-				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					p.logger.Errorln("Received unexpected object: %v", o)
-					return
-				}
-				pod, ok = deletedState.Obj.(*apiv1.Pod)
-				if !ok {
-					p.logger.Errorln("DeletedFinalStateUnknown contained non-Pod object: %v", deletedState.Obj)
-					return
-				}
+			pod, err := convertToPod(o)
+			if err != nil {
+				p.logger.With("err", err).Errorln("converting to Pod object failed")
+				return
 			}
-
 			send(&config.TargetGroup{Source: podSource(pod)})
 		},
 		UpdateFunc: func(_, o interface{}) {
-			send(p.buildPod(o.(*apiv1.Pod)))
+			pod, err := convertToPod(o)
+			if err != nil {
+				p.logger.With("err", err).Errorln("converting to Pod object failed")
+				return
+			}
+			send(p.buildPod(pod))
 		},
 	})
 
 	// Block until the target provider is explicitly canceled.
 	<-ctx.Done()
+}
+
+func convertToPod(o interface{}) (*apiv1.Pod, error) {
+	pod, isPod := o.(*apiv1.Pod)
+	if !isPod {
+		deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("Received unexpected object: %v", o)
+		}
+		pod, ok = deletedState.Obj.(*apiv1.Pod)
+		if !ok {
+			return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Pod object: %v", deletedState.Obj)
+		}
+	}
+
+	return pod, nil
 }
 
 const (

--- a/retrieval/discovery/kubernetes/pod.go
+++ b/retrieval/discovery/kubernetes/pod.go
@@ -74,7 +74,21 @@ func (p *Pod) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 			send(p.buildPod(o.(*apiv1.Pod)))
 		},
 		DeleteFunc: func(o interface{}) {
-			send(&config.TargetGroup{Source: podSource(o.(*apiv1.Pod))})
+			pod, isPod := o.(*apiv1.Pod)
+			if !isPod {
+				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					p.logger.Errorln("Received unexpected object: %v", o)
+					return
+				}
+				pod, ok = deletedState.Obj.(*apiv1.Pod)
+				if !ok {
+					p.logger.Errorln("DeletedFinalStateUnknown contained non-Pod object: %v", deletedState.Obj)
+					return
+				}
+			}
+
+			send(&config.TargetGroup{Source: podSource(pod)})
 		},
 		UpdateFunc: func(_, o interface{}) {
 			send(p.buildPod(o.(*apiv1.Pod)))

--- a/retrieval/discovery/kubernetes/pod_test.go
+++ b/retrieval/discovery/kubernetes/pod_test.go
@@ -14,13 +14,13 @@
 package kubernetes
 
 import (
-	//"fmt"
 	"testing"
 
 	"github.com/prometheus/common/log"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
 	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/tools/cache"
 )
 
 func podStoreKeyFunc(obj interface{}) (string, error) {
@@ -196,6 +196,43 @@ func TestPodDiscoveryDelete(t *testing.T) {
 	k8sDiscoveryTest{
 		discovery:  n,
 		afterStart: func() { go func() { i.Delete(makePod()) }() },
+		expectedInitial: []*config.TargetGroup{
+			&config.TargetGroup{
+				Targets: []model.LabelSet{
+					model.LabelSet{
+						"__address__":                                   "1.2.3.4:9000",
+						"__meta_kubernetes_pod_container_name":          "testcontainer",
+						"__meta_kubernetes_pod_container_port_name":     "testport",
+						"__meta_kubernetes_pod_container_port_number":   "9000",
+						"__meta_kubernetes_pod_container_port_protocol": "TCP",
+					},
+				},
+				Labels: model.LabelSet{
+					"__meta_kubernetes_pod_name":      "testpod",
+					"__meta_kubernetes_namespace":     "default",
+					"__meta_kubernetes_pod_node_name": "testnode",
+					"__meta_kubernetes_pod_ip":        "1.2.3.4",
+					"__meta_kubernetes_pod_host_ip":   "2.3.4.5",
+					"__meta_kubernetes_pod_ready":     "true",
+				},
+				Source: "pod/default/testpod",
+			},
+		},
+		expectedRes: []*config.TargetGroup{
+			&config.TargetGroup{
+				Source: "pod/default/testpod",
+			},
+		},
+	}.Run(t)
+}
+
+func TestPodDiscoveryDeleteUnknownCacheState(t *testing.T) {
+	n, i := makeTestPodDiscovery()
+	i.GetStore().Add(makePod())
+
+	k8sDiscoveryTest{
+		discovery:  n,
+		afterStart: func() { go func() { i.Delete(cache.DeletedFinalStateUnknown{Obj: makePod()}) }() },
 		expectedInitial: []*config.TargetGroup{
 			&config.TargetGroup{
 				Targets: []model.LabelSet{

--- a/retrieval/discovery/kubernetes/service.go
+++ b/retrieval/discovery/kubernetes/service.go
@@ -14,6 +14,7 @@
 package kubernetes
 
 import (
+	"fmt"
 	"net"
 	"strconv"
 
@@ -61,32 +62,49 @@ func (s *Service) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 	}
 	s.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(o interface{}) {
-			send(s.buildService(o.(*apiv1.Service)))
+			svc, err := convertToService(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Service object failed")
+				return
+			}
+			send(s.buildService(svc))
 		},
 		DeleteFunc: func(o interface{}) {
-			service, isService := o.(*apiv1.Service)
-			if !isService {
-				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					s.logger.Errorln("Received unexpected object: %v", o)
-					return
-				}
-				service, ok = deletedState.Obj.(*apiv1.Service)
-				if !ok {
-					s.logger.Errorln("DeletedFinalStateUnknown contained non-Service object: %v", deletedState.Obj)
-					return
-				}
+			svc, err := convertToService(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Service object failed")
+				return
 			}
-
-			send(&config.TargetGroup{Source: serviceSource(service)})
+			send(&config.TargetGroup{Source: serviceSource(svc)})
 		},
 		UpdateFunc: func(_, o interface{}) {
-			send(s.buildService(o.(*apiv1.Service)))
+			svc, err := convertToService(o)
+			if err != nil {
+				s.logger.With("err", err).Errorln("converting to Service object failed")
+				return
+			}
+			send(s.buildService(svc))
 		},
 	})
 
 	// Block until the target provider is explicitly canceled.
 	<-ctx.Done()
+}
+
+func convertToService(o interface{}) (*apiv1.Service, error) {
+	service, isService := o.(*apiv1.Service)
+	if !isService {
+		deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return nil, fmt.Errorf("Received unexpected object: %v", o)
+		}
+		service, ok = deletedState.Obj.(*apiv1.Service)
+		if !ok {
+			return nil, fmt.Errorf("DeletedFinalStateUnknown contained non-Service object: %v", deletedState.Obj)
+		}
+	}
+
+	return service, nil
 }
 
 func serviceSource(s *apiv1.Service) string {

--- a/retrieval/discovery/kubernetes/service.go
+++ b/retrieval/discovery/kubernetes/service.go
@@ -64,7 +64,21 @@ func (s *Service) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
 			send(s.buildService(o.(*apiv1.Service)))
 		},
 		DeleteFunc: func(o interface{}) {
-			send(&config.TargetGroup{Source: serviceSource(o.(*apiv1.Service))})
+			service, isService := o.(*apiv1.Service)
+			if !isService {
+				deletedState, ok := o.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					s.logger.Errorln("Received unexpected object: %v", o)
+					return
+				}
+				service, ok = deletedState.Obj.(*apiv1.Service)
+				if !ok {
+					s.logger.Errorln("DeletedFinalStateUnknown contained non-Service object: %v", deletedState.Obj)
+					return
+				}
+			}
+
+			send(&config.TargetGroup{Source: serviceSource(service)})
 		},
 		UpdateFunc: func(_, o interface{}) {
 			send(s.buildService(o.(*apiv1.Service)))

--- a/retrieval/discovery/kubernetes/service_test.go
+++ b/retrieval/discovery/kubernetes/service_test.go
@@ -21,6 +21,7 @@ import (
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
 	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/tools/cache"
 )
 
 func serviceStoreKeyFunc(obj interface{}) (string, error) {
@@ -147,6 +148,37 @@ func TestServiceDiscoveryDelete(t *testing.T) {
 	k8sDiscoveryTest{
 		discovery:  n,
 		afterStart: func() { go func() { i.Delete(makeService()) }() },
+		expectedInitial: []*config.TargetGroup{
+			&config.TargetGroup{
+				Targets: []model.LabelSet{
+					model.LabelSet{
+						"__meta_kubernetes_service_port_protocol": "TCP",
+						"__address__":                             "testservice.default.svc:30900",
+						"__meta_kubernetes_service_port_name":     "testport",
+					},
+				},
+				Labels: model.LabelSet{
+					"__meta_kubernetes_service_name": "testservice",
+					"__meta_kubernetes_namespace":    "default",
+				},
+				Source: "svc/default/testservice",
+			},
+		},
+		expectedRes: []*config.TargetGroup{
+			&config.TargetGroup{
+				Source: "svc/default/testservice",
+			},
+		},
+	}.Run(t)
+}
+
+func TestServiceDiscoveryDeleteUnknownCacheState(t *testing.T) {
+	n, i := makeTestServiceDiscovery()
+	i.GetStore().Add(makeService())
+
+	k8sDiscoveryTest{
+		discovery:  n,
+		afterStart: func() { go func() { i.Delete(cache.DeletedFinalStateUnknown{Obj: makeService()}) }() },
 		expectedInitial: []*config.TargetGroup{
 			&config.TargetGroup{
 				Targets: []model.LabelSet{

